### PR TITLE
{devel}[GCCcore/12.3.0] Qt6 v6.5.2

### PR DESCRIPTION
**Pull request by appolloford**
(created using `eb --new-pr`)

### CODE REVIEW

File: easybuild/easyconfigs/q/Qt6/Qt6-6.5.2-GCCcore-12.3.0.eb [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+#       |           ^~~
+
+preconfigopts = 'sed -i "s/MultiMedia/Multimedia/g" ../qt-everywhere-src-%(version)s/qtcharts/CMakeLists.txt &&'  # Typo
+configopts = '-Wno-dev '
+configopts += '-DFEATURE_qtpdf_build=OFF '  # Requires CUPS
+configopts += '-DBUILD_qtwayland=OFF '  # Does not work on CentOS 7
+# configopts += '-DBUILD_qtgamepad=OFF '  # Gamepad does not work on CentOS 7. It have been removed from Qt6.0.0 and has not been added back
+
+# make sure QtWebEngine component is being built & installed
+#check_qtwebengine = True
```
Comment (smoors) [inline easybuild/easyconfigs/q/Qt6/Qt6-6.5.2-GCCcore-12.3.0.eb:100]:
```suggestion
# check_qtwebengine = True
```

**Comment by verdurin** [comment]
That's actually a weird failure with `LLVM`:
```
-- Looking for __x86_64__
-- Looking for __x86_64__ - found
-- Found Git: /bin/git (found version "1.8.3.1")
-- Targeting X86
CMake Error at CMakeLists.txt:1171 (add_subdirectory):
  add_subdirectory given source
  "/dev/shm/LLVM/16.0.6/GCCcore-12.3.0/llvm-16.0.6.src/../third-party/unittest"
  which is not an existing directory.


-- Registering ExampleIRTransforms as a pass plugin (static build: OFF)
-- Registering Bye as a pass plugin (static build: OFF)
-- Configuring incomplete, errors occurred!
 (at easybuild/tools/run.py:643 in parse_cmd_output)
 ```

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 8 out of 9 (1 easyconfigs in total)
bear-pg0105u03b.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), Python 3.6.8
See https://gist.github.com/branfosj/2c73905acc7d4b2065768bdb6a4ec4e6 for a full test report.

**Comment by verdurin** [comment]
> That's actually a weird failure with `LLVM`:
> 
> ```
> -- Looking for __x86_64__
> -- Looking for __x86_64__ - found
> -- Found Git: /bin/git (found version "1.8.3.1")
> -- Targeting X86
> CMake Error at CMakeLists.txt:1171 (add_subdirectory):
>   add_subdirectory given source
>   "/dev/shm/LLVM/16.0.6/GCCcore-12.3.0/llvm-16.0.6.src/../third-party/unittest"
>   which is not an existing directory.
> 
> 
> -- Registering ExampleIRTransforms as a pass plugin (static build: OFF)
> -- Registering Bye as a pass plugin (static build: OFF)
> -- Configuring incomplete, errors occurred!
>  (at easybuild/tools/run.py:643 in parse_cmd_output)
> ```

Re-trying now with the easy block required for LLVM.

**Comment by branfosj** [comment]
@boegelbot please test @ generoso
CORE_CNT=16

**Comment by verdurin** [comment]
Test report by @verdurin
**FAILED**
Build succeeded for 0 out of 3 (1 easyconfigs in total)
easybuild-c7.novalocal - Linux CentOS Linux 7.9.2009, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.6.8
See https://gist.github.com/verdurin/cb4a26cb2a6545dce655de3d9845e78f for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 1 out of 2 (1 easyconfigs in total)
bear-pg0211u03a.bear.cluster - Linux Ubuntu 20.04.2 LTS (Focal Fossa), x86_64, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz (cascadelake), Python 3.8.5
See https://gist.github.com/branfosj/06ef3ae2a7ec492f241f40c42028884d for a full test report.

File: easybuild/easyconfigs/q/Qt6/Qt6-6.5.2-GCCcore-12.3.0.eb [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    ('libwebp', '1.3.1'),
+    ('JasPer', '4.0.0'),
+    ('HarfBuzz', '5.3.1'),
+    ('SQLite', '3.42.0'),
+    ('graphite2', '1.3.14'),
+    ('assimp', '5.2.5'),  # for Qt 3D
+    ('FFmpeg', '6.0'),
+    ('X11', '20230603'),
+    ('fontconfig', '2.14.2'),
+    ('html5lib', '1.1'),  # WebEngine
```
Comment (branfosj) [inline easybuild/easyconfigs/q/Qt6/Qt6-6.5.2-GCCcore-12.3.0.eb:64]:
```suggestion
    ('Python', '3.11.3'),
    ('Python-bundle-PyPI', '2023.06'),  # WebEngine
```

[`Python-bundle-PyPi` contains `html5lb`](https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/p/Python-bundle-PyPI/Python-bundle-PyPI-2023.06-GCCcore-12.3.0.eb#L398), so we should use that.

File: easybuild/easyconfigs/q/Qt6/Qt6-6.5.2-GCCcore-12.3.0.eb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+builddependencies = [
+    ('binutils', '2.40'),
+    ('pkgconf', '1.9.5'),
+    ('CMake', '3.26.3'),
+    ('Ninja', '1.11.1'),
+    # deps for QtWebEngine
+    ('Bison', '3.8.2'),
+    ('flex', '2.6.4'),
+    ('gperf', '3.1'),
+    ('Python', '3.11.3'),
```
Comment (branfosj) [inline easybuild/easyconfigs/q/Qt6/Qt6-6.5.2-GCCcore-12.3.0.eb:45]:
```suggestion
```

**Comment by branfosj** [comment]
To avoid the CMake configure error I had to add `configopts += '-DQT_AVOID_CMAKE_ARCHIVING_API=ON '`

I also had to add
```
preconfigopts = 'sed -i "23i set(Python3_ROOT_DIR \$ENV{EBROOTPYTHON})" ../qt-everywhere-src-6.5.2/qtwebengine/src/gn/CMakeLists.txt &&'
build_cmd = "cmake --build . --parallel"
install_cmd = "cmake --install . #"
```

The `preconfigopts` was needed to get the correct Python3 detected during the CMake configure of that component, which happens during the build stage.

The `build_cmd` and `install_cmd` are not the correct way of doing those. But `make` failed for me (`make: *** No targets specified and no makefile found.  Stop.`).

**Comment by Micket** [comment]
@branfosj It's a CMakeNinja easyblock though (Qt insists on using Ninja, spits out a warning for CMakeMake). Does `ninja` also fail?

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
zen2-rockylinux-88 - Linux Rocky Linux 8.8, x86_64, AMD EPYC 7452 32-Core Processor (zen2), Python 3.6.8
See https://gist.github.com/SebastianAchilles/2147f1f53db3d66c5ce61f119aa47e03 for a full test report.

**Comment by branfosj** [comment]
> @branfosj It's a CMakeNinja easyblock though (Qt insists on using Ninja, spits out a warning for CMakeMake). Does `ninja` also fail?

https://github.com/easybuilders/easybuild-easyblocks/pull/2993 will fix the issue with `make` being used.

**Comment by branfosj** [comment]
Test report by @branfosj
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2993
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0105u03a.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), Python 3.6.8
See https://gist.github.com/branfosj/b4c3caf1f08263b0a4c9ffe908e175f0 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2993
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0211u03a.bear.cluster - Linux Ubuntu 20.04.2 LTS (Focal Fossa), x86_64, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz (cascadelake), Python 3.8.5
See https://gist.github.com/branfosj/1189ffa740d442abb049aff1d5469ef7 for a full test report.

File: easybuild/easyconfigs/q/Qt6/Qt6-6.5.2_fix_OF-Gentoo.patch [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+author: Kenneth Hoste (HPC-UGent)
+diff --git a/qt3d/src/3rdparty/assimp/src/contrib/zlib/contrib/minizip/ioapi.h b/qt3d/src/3rdparty/assimp/src/contrib/zlib/contrib/minizip/ioapi.h
+index 8dcbdb06e3..e9cc96914e 100644
+--- a/qt3d/src/3rdparty/assimp/src/contrib/zlib/contrib/minizip/ioapi.h
++++ b/qt3d/src/3rdparty/assimp/src/contrib/zlib/contrib/minizip/ioapi.h
+@@ -130,6 +130,14 @@ extern "C" {
+ #endif
+
+
++// if OF macro is not defined by _Z_OF is, define OF as _Z_OF
```
Comment (akesandgren) [inline easybuild/easyconfigs/q/Qt6/Qt6-6.5.2_fix_OF-Gentoo.patch:18]: Should probably be "if OF macro is not defined but _Z_OF is"

And the same for the one below.

File: easybuild/easyconfigs/q/Qt6/Qt6-6.5.2-GCCcore-12.3.0.eb [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+preconfigopts = 'sed -i "s/MultiMedia/Multimedia/g" ../qt-everywhere-src-%(version)s/qtcharts/CMakeLists.txt &&'  # Typo
+preconfigopts += 'sed -i "23i set(Python3_ROOT_DIR \\$ENV{EBROOTPYTHON})" ' \
+                 '../qt-everywhere-src-6.5.2/qtwebengine/src/gn/CMakeLists.txt &&'
+
+configopts = '-Wno-dev '
+configopts += '-DFEATURE_qtpdf_build=OFF '  # Requires CUPS
+configopts += '-DQT_AVOID_CMAKE_ARCHIVING_API=ON '
+configopts += '-DPython3_ROOT_DIR=$EBROOTPYTHON '
+configopts += '-DBUILD_qtwayland=OFF '  # Does not work on CentOS 7
+# Removed from Qt6.0.0 by may be added back in the future
```
Comment (akesandgren) [inline easybuild/easyconfigs/q/Qt6/Qt6-6.5.2-GCCcore-12.3.0.eb:84]: This should also be "Removed from Qt6.0.0 but may be"

**Review by akesandgren — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/q/Qt6/Qt6-6.5.2_fix_OF-Gentoo.patch (hunk 4 above)
Comment (appolloford) [reply easybuild/easyconfigs/q/Qt6/Qt6-6.5.2_fix_OF-Gentoo.patch:18]: Yeah it should be. I copied this paragraph from the older patch but did not check

**Comment by akesandgren** [comment]
Test report by @akesandgren
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
b-cn1603.hpc2n.umu.se - Linux Ubuntu 22.04, x86_64, AMD EPYC 7313 16-Core Processor, 1 x NVIDIA NVIDIA A100 80GB PCIe, 525.125.06, Python 3.10.12
See https://gist.github.com/akesandgren/c69fe874a9ac9e666547ce56ddfe11bf for a full test report.

File: easybuild/easyconfigs/q/Qt6/Qt6-6.5.2_fix_OF-Gentoo.patch (hunk 4 above)
Comment (akesandgren) [reply easybuild/easyconfigs/q/Qt6/Qt6-6.5.2_fix_OF-Gentoo.patch:18]: Can you please make a commit for these "by" -> "but" soon. I'd like to get this one merged

**Review by akesandgren — DISMISSED** [review]
LGTM

File: easybuild/easyconfigs/q/Qt6/Qt6-6.5.2-GCCcore-12.3.0.eb [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    'https://download.qt.io/archive/qt/%(version_major_minor)s/%(version)s/single/',
+    'https://download.qt.io/new_archive/qt/%(version_major_minor)s/%(version)s/single/',
+]
+sources = ['qt-everywhere-src-%(version)s.tar.xz']
+patches = [
+    # they may not be needed if qmake is not used?
+    # QT_CFLAGS_ARCH_HASWELL in cmake/QtCompilerOptimization.cmake should be patched?
+    # 'Qt5-5.13.1_fix-avx2.patch',
+    # 'Qt5-5.13.1_fix-qmake-libdir.patch',
+    # 'Qt5-5.15.2_fix-qtwebengine-breakpad-glibc-2.34.patch',
```
Comment (appolloford) [inline easybuild/easyconfigs/q/Qt6/Qt6-6.5.2-GCCcore-12.3.0.eb:24]: Looks these patches are not needed anymore. I may remove them in the next commit.

**Review by akesandgren — APPROVED** [review]
LGTM

**Comment by akesandgren** [comment]
Going in, thanks @appolloford!

**Pull request merged**